domain/state.go

```diff
@@ -99,6 +100,84 @@ func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statemen
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting database")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// The txContext is created with the transaction and passed to the
+		// closure function. This ensures that the transaction is always
+		// available to the closure. Once the transaction is complete, the
+		// transaction is removed from the txContext. This is to prevent
+		// the transaction from being used outside of the transaction scope.
+		// This will prevent any references to the sqlair.TX from being held
+		// outside of the transaction scope.
+
+		// TODO (stickupkid): The txContext can be pooled on the StateBase to
+		// reduce the number of allocations. Attempting to push the tx into
+		// the context would prevent that as a viable option.
+		txCtx := &txContext{
+			Context: ctx,
+			t:       tx,
+		}
+		defer txCtx.removeTx()
+
+		return fn(txCtx)
+	})
+}
+
+// TxStateBase is an interface that provides a method for executing a closure
+// within the scope of a transaction.
+type TxStateBase interface {
+	// RunTx executes the closure function within the scope of a transaction.
+	// The closure is passed a TxContext that can be passed on to state
+	// functions, so that they can perform work within that same transaction.
+	// The closure will be retried according to the transaction retry semantics,
+	// if the transaction fails due to transient errors.
+	// The closure should only be used to perform state changes and must not be
+	// used to execute queries outside of the state scope.
+	RunTx(ctx context.Context, fn func(TxContext) error) error
+}
+
+// Run executes the closure function using the provided TxContext as
+// the transaction context.
+// It is expected that the closure will perform state changes within the
+// transaction scope.
+// Any errors returned from the closure are coerced into a standard error
+// to prevent sqlair errors from being returned to the Service layer.
+func Run(ctx TxContext, fn func(context.Context, *sqlair.TX) error) error {
+	txCtx, ok := ctx.(*txContext)
+	if !ok {
+		// If you're seeing this error, it means that the TxContext was not
+		// created by RunTx. This is a programming error. Did you attempt to
+		// wrap the context in a custom context and pass it to Run?
+		return fmt.Errorf("programmatic error: TxContext is not a *txContext: %T", ctx)
+	}
+
+	tx := txCtx.tx()
+	if tx == nil {
+		// If you're seeing this error, it means that the TxContext was not
+		// created by RunTx. This is a programming error. Did you capture
+		// the TxContext from a RunTx closure and try to use it outside of
+		// the closure?
+		return fmt.Errorf("programmatic error: TxContext does not have a transaction")
+	}
+
+	// Execute the function with the transaction.
+	// Coerce the error to ensure that no sql or sqlair errors are returned
+	// from the function and into the Service layer.
+	return CoerceError(fn(ctx, tx))
```

Comment:
We must hold the mutex for the life of the call to `fn` closure. Essentially, while the `sqlair.Tx` is loaned out, it must be locked until it is "returned" to `txContext`.